### PR TITLE
Make toxcore code C++ compatible.

### PR DESCRIPTION
**Pull request by iphydf**
It is still C code, so still compatible with C compilers as well. This
change lets us see more clearly where implicit conversions occur by
making them explicit.

---
This change is NOT reviewable. Use Github reviews instead.

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/toxcore/143)

### CODE REVIEW

**Comment by nbraud** [comment]
Reviewed 45 of 45 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[auto_tests/dht_test.c, line 531 at r1](https://reviewable.io:443/reviews/toktok/toxcore/143#-KRuwhgdu5AZ7wA6xHZi:-KRuwhgdu5AZ7wA6xHZj:bulajjl) ([raw file](https://github.com/toktok/toxcore/blob/1d8755f98d557ff9af057d8e6603a351c6fd564f/auto_tests/dht_test.c#L531)):*
> ```C
> 
>                 ck_assert_msg(count == 1, "Nodes in search don't know ip of friend. %u %u %u", i, j, count);
> #if 0
> ```

Why is `#if 0` required for C++ compat?

---

*[other/bootstrap_daemon/src/config.c, line 367 at r1](https://reviewable.io:443/reviews/toktok/toxcore/143#-KRuxior-8qxWa2CFtp5:-KRuxior-8qxWa2CFtp6:b-if8hf3) ([raw file](https://github.com/toktok/toxcore/blob/1d8755f98d557ff9af057d8e6603a351c6fd564f/other/bootstrap_daemon/src/config.c#L367)):*
> ```C
> 
>     while (config_setting_length(node_list)) {
>         int address_resolved;
> ```

Any specific reason for moving the definition site (and losing `cons`)?

---

*[toxcore/list.c, line 116 at r1](https://reviewable.io:443/reviews/toktok/toxcore/143#-KRv-EiNwPKG3hj95LX7:-KRv-EiOxILw82N5RH4C:b-i434hf) ([raw file](https://github.com/toktok/toxcore/blob/1d8755f98d557ff9af057d8e6603a351c6fd564f/toxcore/list.c#L116)):*
> ```C
> static int resize(BS_LIST *list, uint32_t new_size)
> {
>     uint8_t *data = (uint8_t *)realloc(list->data, list->element_size * new_size);
> ```

The pointer could be `const`, same for `ids`.

---

*[toxcore/Messenger.c, line 2532 at r1](https://reviewable.io:443/reviews/toktok/toxcore/143#-KRv-UQugZdpL5Zvc4dl:-KRv-UQvzPnDwLJPy2JN:bpsjqfj) ([raw file](https://github.com/toktok/toxcore/blob/1d8755f98d557ff9af057d8e6603a351c6fd564f/toxcore/Messenger.c#L2532)):*
> ```C
> 
> 
>         uint32_t myfriend, dhtfriend;
> ```

It's strange for `myfriend` to be a `uint32`.
Perhaps `friend_idx`?

---

*[toxcore/TCP_connection.c, line 41 at r1](https://reviewable.io:443/reviews/toktok/toxcore/143#-KRv0Dk_mU8Tmq1QYCBn:-KRv0DkaXhVO5Zu4zIOO:b-jgv2ce) ([raw file](https://github.com/toktok/toxcore/blob/1d8755f98d557ff9af057d8e6603a351c6fd564f/toxcore/TCP_connection.c#L41)):*
> ```C
>         ? (temp_pointer = (element_type *)realloc( \
>               array, \
>               (num) * sizeof(element_type)), \
> ```

This can overflow, and such a macro would actually be a nice place to catch it.
(Obviously out of scope for this PR)

---

*[toxcore/tox.c, line 845 at r1](https://reviewable.io:443/reviews/toktok/toxcore/143#-KRv1AQE9EL_fnLazHJW:-KRv1AQF8qW9hw_ipbV_:b-lqhjmt) ([raw file](https://github.com/toktok/toxcore/blob/1d8755f98d557ff9af057d8e6603a351c6fd564f/toxcore/tox.c#L845)):*
> ```C
>     if (ret == USERSTATUS_INVALID) {
>         SET_ERROR_PARAMETER(error, TOX_ERR_FRIEND_QUERY_FRIEND_NOT_FOUND);
>         return (TOX_USER_STATUS)(TOX_USER_STATUS_BUSY + 1);
> ```

If the convention is to return an invalid enum value, I think there should be a macro defined for it (rather than risk pandemonium once another enum value gets added).

---

*[toxcore/tox.c, line 1352 at r1](https://reviewable.io:443/reviews/toktok/toxcore/143#-KRv1zOUW_P8nMOYm4N-:-KRv1zOVvW_huybJfcMO:b-4dnakd) ([raw file](https://github.com/toktok/toxcore/blob/1d8755f98d557ff9af057d8e6603a351c6fd564f/toxcore/tox.c#L1352)):*
> ```C
> }
> 
> /**********GROUP CHAT FUNCTIONS: WARNING Group chats will be rewritten so this might change ************/
> ```

Is moving the code from `tox_group.c` needed for C++ compat?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/toxcore/143)*

**Comment by iphydf** [comment]
I removed reviewable, so I'm going to reply in here:
- Why is if 0 required? It's not, that's just a drive-by cleanup. `#if 0` is preferred to comments to comment out blocks of code in C. The main reason is that code formatters can see them, a second reason is that enabling it is a single bit change in the source code.
- Why move the definition and lose `const`? Because `goto` jumps over the initialisation, which is ill-formed in C++.
- Pointers should be `const`. That's not required for C++, and none of the other code does this. Perhaps we can do a transformation over the code at some point and make all non-changing names `const`.
- `myfriend` to `friend_idx`. Done.
- Overflow: I don't know if it actually can, but yes, by itself that could happen. Toxcore code tends not to check for this, because "we know that it can't overflow because we have 30% of the code in our heads at all times" is how most of the code is supposed to be read. We'll be looking into those things at some point.
- Returning invalid values: it should not return an invalid value. The value should be unspecified, and no code should rely on it being something known, but it should return valid values. Out of scope for this PR though.
- Moving code: no, not required, but another PR was going to do that, and I wanted this PR to be compatible with it. That PR is merged now.

File: toxcore/DHT.c [hunk 1]
```diff
@@ -985,24 +985,24 @@ static unsigned int ping_node_from_getnodes_ok(DHT *dht, const uint8_t *public_k
     for (i = 0; i < dht->num_friends; ++i) {
         bool store_ok = 0;
 
-        DHT_Friend *friend = &dht->friends_list[i];
+        DHT_Friend *dht_friend = &dht->friends_list[i];
```
Comment (nbraud) [inline toxcore/DHT.c]: @nurupo Yes, `friend` is a C++ keyword.

**Comment by nbraud** [comment]
@iphydf OK for me, I will open issues for the out-of-scope comments

File: toxcore/DHT.c (hunk 1 above)
Comment (GrayHatter) [inline toxcore/DHT.c]: Again, yes. Friend in the context of DHT is problematic.

Breaking merges or otherwise this is a positive change.

**Pull request merged**